signature/src/agent.rs

```diff
@@ -0,0 +1,145 @@
+#[derive(Serialize, Deserialize, Debug)]
+struct ResourceDescription {
+    name: String,
+    optional: HashMap<String, String>,
+}
+
+impl ResourceDescription {
+    /// Create a new ResourceDescription with resource name.
+    pub fn new(name: &str) -> Self {
+        ResourceDescription {
+            name: name.to_string(),
+            optional: HashMap::new(),
+        }
+    }
+}
+
+impl Agent {
+    /// Create a new signature-agent, the input parameter:
+    /// * `aa_kbc_params`: s string with format `<kbc_name>::<kbs_uri>`.
+    pub async fn new(aa_kbc_params: &str) -> Result<Self> {
+        // unzip here is unstable
+        if let Some((kbc_name, kbs_uri)) = aa_kbc_params.split_once("::") {
+            if kbc_name.is_empty() {
+                return Err(anyhow!("aa_kbc_params: missing KBC name"));
+            }
+
+            if kbs_uri.is_empty() {
+                return Err(anyhow!("aa_kbc_params: missing KBS URI"));
+            }
+
+            Ok(Self {
+                client: GetResourceServiceClient::connect(AA_GETRESOURCE_ADDR).await?,
+                kbc_name: kbc_name.into(),
+                kbc_uri: kbs_uri.into(),
+            })
+        } else {
+            Err(anyhow!("aa_kbc_params: KBC/KBS pair not found"))
+        }
+    }
+
+    /// Get resource from using, using `resource_name` as `name` in a ResourceDescription.
+    /// Please refer to https://github.com/confidential-containers/image-rs/blob/main/docs/ccv1_image_security_design.md#get-resource-service
+    /// for more information.
+    /// Then save the gathered data into `path`
+    async fn get_resource(&mut self, resource_name: &str, path: &str) -> Result<()> {
+        let resource_description = serde_json::to_string(&ResourceDescription::new(resource_name))?;
+        let req = tonic::Request::new(GetResourceRequest {
+            kbc_name: self.kbc_name.clone(),
+            kbs_uri: self.kbc_uri.clone(),
+            resource_description,
+        });
+        let res = self.client.get_resource(req).await?;
+
+        fs::write(path, res.into_inner().resource).await?;
+        Ok(())
+    }
+
+    /// `check_policy` judges whether the container image is allowed to be pulled and run.
+    ///
+    /// According to the configuration of the policy file. The policy may include
+    /// signatures, if a signature of the container image needs to be verified, the
+    /// specified signature scheme is used for signature verification.
+    pub async fn allows_image(&mut self, image_reference: &str, image_digest: &str) -> Result<()> {
+        if !Path::new(IMAGE_SECURITY_CONFIG_DIR).exists() {
+            fs::create_dir_all(IMAGE_SECURITY_CONFIG_DIR)
+                .await
+                .map_err(|e| anyhow!("Create image security runtime config dir failed: {:?}", e))?;
+        }
+
+        // if Policy config file does not exist, get if from KBS.
+        if !Path::new(POLICY_FILE_PATH).exists() {
```

Comment:
Maybe your question here is whether we need to check the file `POLICY_FILE_PATH` every time we call `allows_image`. However, there are two places using AA to access KBC:
* Every time we verify a signature and need to get resource
* Get the policy file

In the original design, every time the main entrance `pull_image()` is called by `kata-agent`, the check for policy's existence will be done. Now I put the check inside `agent` for signature verification, it still follows that every time the main entrance of `image-rs` is called, the check will be done once.

Some optimizations can be taken to avoid useless check, but IMO it will be in the initialization of `kata-agent`, thus in the whole lifetime of a `kata-agent`, the check will only be done once.